tools/plugins/com.liferay.ide.upgrade.core/src/com/liferay/blade/upgrade/liferay70/apichanges/BaseLiferayDescriptorVersion.java

```diff
@@ -0,0 +1,131 @@
+import com.liferay.blade.api.AutoMigrator;
+import com.liferay.blade.api.Problem;
+import com.liferay.blade.api.SearchResult;
+import com.liferay.blade.api.XMLFile;
+import com.liferay.blade.upgrade.XMLFileMigrator;
+
+import java.io.File;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.wst.sse.core.StructuredModelManager;
+import org.eclipse.wst.sse.core.internal.provisional.IModelManager;
+import org.eclipse.wst.xml.core.internal.document.DocumentTypeImpl;
+import org.eclipse.wst.xml.core.internal.provisional.document.IDOMDocument;
+import org.eclipse.wst.xml.core.internal.provisional.document.IDOMModel;
+
+/**
+ * @author Seiphon Wang
+ */
+@SuppressWarnings("restriction")
+public abstract class BaseLiferayDescriptorVersion extends XMLFileMigrator implements AutoMigrator {
+
+	public BaseLiferayDescriptorVersion(String publicIDPattern, String version) {
+		_publicIDPattern = publicIDPattern;
+		_version = version;
+	}
+
+	@Override
+	public int correctProblems(File file, List<Problem> problems) throws AutoMigrateException {
+		try {
+			IFile xmlFile = getXmlFile(file);
+
+			IModelManager modelManager = StructuredModelManager.getModelManager();
+
+			IDOMModel domModel = (IDOMModel)modelManager.getModelForRead(xmlFile);
+
+			IDOMDocument document = domModel.getDocument();
+
+			DocumentTypeImpl docType = (DocumentTypeImpl)document.getDoctype();
+
+			int problemsFixed = 0;
+
+			for (int i = 0; i < problems.size(); i++) {
+				if (docType != null) {
+					String publicId = docType.getPublicId();
+
+					String newPublicId = _getNewDoctTypeSetting(publicId, _version, _PUBLICID_REGREX);
+
+					docType.setPublicId(newPublicId);
+
+					String systemId = docType.getSystemId();
+
+					String newSystemId = _getNewDoctTypeSetting(
+						systemId, _version.replaceAll("\\.", "_"), _SYSTEMID_REGREX);
+
+					docType.setSystemId(newSystemId);
+
+					problemsFixed++;
+				}
+
+				domModel.save();
+			}
+
+			return problemsFixed;
+		}
+		catch (Exception e) {
+		}
+
+		return 0;
+	}
+
+	@Override
+	protected List<SearchResult> searchFile(File file, XMLFile xmlFileChecker) {
+		List<SearchResult> results = new ArrayList<>();
+
+		for (String liferayDtdName : _liferayDtdNames) {
+			results.add(xmlFileChecker.findDocumentTypeDeclaration(liferayDtdName, _publicIDPattern));
+		}
+
+		return results;
+	}
+
+	private String _getNewDoctTypeSetting(String doctypeSetting, String newValue, String regrex) {
+		String newDoctTypeSetting = null;
+
+		Pattern p = Pattern.compile(regrex, Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+
+		Matcher m = p.matcher(doctypeSetting);
+
+		if (m.find()) {
+			String oldVersionString = m.group(m.groupCount());
+
+			newDoctTypeSetting = doctypeSetting.replace(oldVersionString, newValue);
+		}
+
+		return newDoctTypeSetting;
+	}
+
+	private static final String _PUBLICID_REGREX =
```

Comment:
@Seiphon  We should save the compiled Pattern.  Don't just save the strings, always do
`private static final Pattern _PUBLICID_PATTERN = Pattern.compile("", options..);`
cc @jtydhr88